### PR TITLE
handle errors uploading files more gracefully

### LINKED ISSUE #6029: amazon_ebs panic: runtime error: invalid memory address or nil pointer dereference
Using `packer -v 1.2.1` on `amazon linux`.

Part of crash log...
```
2018/03/19 15:23:23 ui: ESC[1;32m==> amazon-ebs: Connected to SSH!ESC[0m
2018/03/19 15:23:23 packer: 2018/03/19 15:23:23 Running the provision hook
2018/03/19 15:23:23 [INFO] (telemetry) Starting provisioner shell
2018/03/19 15:23:23 ui: ESC[1;32m==> amazon-ebs: Provisioning with shell script: /tmp/packer-shell735442832ESC[0m
2018/03/19 15:23:23 packer: 2018/03/19 15:23:23 Opening /tmp/packer-shell735442832 for reading
2018/03/19 15:23:23 packer: 2018/03/19 15:23:23 [INFO] 178 bytes written for 'uploadData'
2018/03/19 15:23:23 [INFO] 178 bytes written for 'uploadData'
2018/03/19 15:23:23 packer: 2018/03/19 15:23:23 opening new ssh session
2018/03/19 15:23:24 packer: 2018/03/19 15:23:24 starting remote command: test -d "/tmp/script_4845.sh"
2018/03/19 15:23:33 packer: 2018/03/19 15:23:33 path is a directory; copying file into directory.
2018/03/19 15:23:33 packer: panic: runtime error: invalid memory address or nil pointer dereference
2018/03/19 15:23:33 packer: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9f19cc]
2018/03/19 15:23:33 packer:
2018/03/19 15:23:33 packer: goroutine 198 [running]:
2018/03/19 15:23:33 packer: github.com/hashicorp/packer/communicator/ssh.(*comm).scpUploadSession(0xc4206b9a70, 0xc4202adb60, 0x13, 0x22d89c0, 0xc42025b2b0, 0x0, 0x0, 0x22d89c0)
2018/03/19 15:23:33 packer:     /Users/mwhooker/go/src/github.com/hashicorp/packer/communicator/ssh/communicator.go:591 +0x41c
2018/03/19 15:23:33 packer: github.com/hashicorp/packer/communicator/ssh.(*comm).Upload(0xc4206b9a70, 0xc4202adb60, 0x13, 0x22d89c0, 0xc42025b2b0, 0x0, 0x0, 0xc420408000)
2018/03/19 15:23:33 packer:     /Users/mwhooker/go/src/github.com/hashicorp/packer/communicator/ssh/communicator.go:156 +0xcc
2018/03/19 15:23:33 packer: github.com/hashicorp/packer/packer/rpc.(*CommunicatorServer).Upload(0xc420106a80, 0xc420421500, 0xc420502770, 0x0, 0x0)
2018/03/19 15:23:33 packer:     /Users/mwhooker/go/src/github.com/hashicorp/packer/packer/rpc/communicator.go:294 +0x14a
2018/03/19 15:23:33 packer: reflect.Value.call(0xc42065b5c0, 0xc42000cdd0, 0x13, 0x20f465e, 0x4, 0xc420795f18, 0x3, 0x3, 0xc420086200, 0xc4204a6388, ...)
2018/03/19 15:23:33 packer:     /usr/local/go/src/reflect/value.go:447 +0x969
2018/03/19 15:23:33 packer: reflect.Value.Call(0xc42065b5c0, 0xc42000cdd0, 0x13, 0xc420498718, 0x3, 0x3, 0x1, 0x704a4e, 0x1)
2018/03/19 15:23:33 packer:     /usr/local/go/src/reflect/value.go:308 +0xa4
2018/03/19 15:23:33 packer: net/rpc.(*service).call(0xc42065c700, 0xc42017c8c0, 0xc4201824b8, 0xc4201824c0, 0xc420488f00, 0xc420421460, 0x1a404c0, 0xc420421500, 0x16, 0x1a96e40, ...)
2018/03/19 15:23:33 packer:     /usr/local/go/src/net/rpc/server.go:384 +0x14e
2018/03/19 15:23:33 packer: created by net/rpc.(*Server).ServeCodec
2018/03/19 15:23:33 packer:     /usr/local/go/src/net/rpc/server.go:480 +0x43a
2018/03/19 15:23:33 /usr/bin/packer: plugin process exited
2018/03/19 15:23:33 [INFO] (telemetry) ending amazon-ebs
2018/03/19 15:23:33 ui error: ESC[1;31mBuild 'amazon-ebs' errored: unexpected EOFESC[0m
2018/03/19 15:23:33 Builds completed. Waiting on interrupt barrier...
2018/03/19 15:23:33 machine readable: error-count []string{"1"}
2018/03/19 15:23:33 ui error:
==> Some builds didn't complete successfully and had errors:
2018/03/19 15:23:33 machine readable: amazon-ebs,error []string{"unexpected EOF"}
2018/03/19 15:23:33 ui error: --> amazon-ebs: unexpected EOF
2018/03/19 15:23:33 ui:
==> Builds finished but no artifacts were created.
2018/03/19 15:23:33 [INFO] (telemetry) Finalizing.
2018/03/19 15:23:33 packer: 2018/03/19 15:23:33 Retryable error: Error uploading script: unexpected EOF
2018/03/19 15:23:33 waiting for all plugin processes to complete...
2018/03/19 15:23:33 [INFO] (telemetry) ending shell
2018/03/19 15:23:33 /usr/bin/packer: plugin process exited
2018/03/19 15:23:33 /usr/bin/packer: plugin process exited
```

packer.json 
```
{
  "variables": {
    "jenkins_job": "{{env `JOB_NAME`}}",
    "jenkins_build": "{{env `BUILD_NUMBER`}}"
  },
  "builders": [{
    "type": "amazon-ebs",
    "ssh_pty" : true,
    "region": "{{user `region`}}",
    "source_ami": "{{user `base_ami`}}",
    "encrypt_boot": true,
    "instance_type": "{{user `instance_type`}}",
    "ssh_username": "{{user `ssh_username`}}",
    "vpc_id": "{{user `vpc_id`}}",
    "subnet_id": "{{user `subnet_id`}}",
    "ami_block_device_mappings": [{
      "device_name": "/dev/xvda",
      "delete_on_termination": "true",
      "volume_size": 50,
      "volume_type": "standard"
    }],
    "tags": {
      "Base Image Id": "{{user `base_ami`}}",
      "Git Commit": "{{user `git_commit`}}",
      "Git Branch": "{{user `git_branch`}}",
      "Git URL": "{{user `git_url`}}",
      "Operating System": "{{user `operating_system`}}",
      "Deprecated": "No",
      "Jenkins Job Name": "{{user `jenkins_job`}}",
      "Jenkins Build Number": "{{user `jenkins_build`}}"
    },
    "ami_name": "{{user `ami_name`}}-{{isotime \"2006.01.02.03.04.05\"}}",
    "ami_users": "{{ user `ami_users` }}",
    "ami_regions": "{{ user `ami_regions` }}",
    "associate_public_ip_address": false
     }],
  "provisioners": [
    {
        "type": "shell",
        "inline": [
            "sleep 30",
            "sudo yum upgrade -y",
            "sudo yum install -y epel-release",
            "sudo yum install -y gcc python-devel libffi-devel gcc openssl-devel zip unzip",
            "sudo pip install ansible"
        ]
    },
    {
      "type": "ansible-local",
      "playbook_file": "playbook.yml",
      "playbook_dir": "./"
    }
  ]
}
```

### LINKED ISSUE #6032: scp uploader assumes we are uploading a file
In a few places in the ssh communicator, we try to access the os.FileInfo object, assuming it will exist. cf https://github.com/hashicorp/packer/blob/b9bcde5a97ab5b936ccb13dbe23f3ee2bd98fa8d/communicator/ssh/communicator.go#L427

This assumption is not correct, because if, for example, we are uploading an inline script, we don't create a file object. We create an object that implements `io.Reader`. The existence of the argument is a bit confusing and would ideally have been avoided, but it's a useful optimization. This is entirely my fault for not catching it in code review, but it's something we need to deal with.

Here is a script that will reproduce it reliably:

```json
{
    "builders": [
        {
            "ami_name": "packer-qs-{{timestamp}}",
            "instance_type": "t2.micro",
            "region": "us-east-1",
            "source_ami_filter": {
                "filters": {
                    "name": "*ubuntu-xenial-16.04-amd64-server-*",
                    "root-device-type": "ebs",
                    "virtualization-type": "hvm"
                },
                "most_recent": true,
                "owners": [
                    "099720109477"
                ]
            },
            "ssh_username": "ubuntu",
            "type": "amazon-ebs"
        }
    ],
    "provisioners": [
        {
            "inline": [
                "sleep 3",
                "echo foo"
            ],
            "remote_path": "/tmp",
            "type": "shell"
        }
    ]
}
```

This may or may not be the same issue as #6029, but they are surely related. Happy to chat more about this face to face

### PR DESCRIPTION
**Pull request by mwhooker**
Fixes #6032, fixes #6029

### CODE REVIEW

**Comment by mwhooker** [comment]
this solves #6029 as well. The output becomes:

```
2018/03/20 11:31:38 ui: ==> amazon-ebs: Connected to SSH!
==> amazon-ebs: Connected to SSH!
2018/03/20 11:31:38 packer: 2018/03/20 11:31:38 Running the provision hook
2018/03/20 11:31:38 [INFO] (telemetry) Starting provisioner shell
2018/03/20 11:31:38 ui: ==> amazon-ebs: Provisioning with shell script: /var/folders/yy/ff0tkhr141xdxvxf5fz43kmw0000gn/T/packer-shell574671846
2018/03/20 11:31:38 packer: 2018/03/20 11:31:38 Opening /var/folders/yy/ff0tkhr141xdxvxf5fz43kmw0000gn/T/packer-shell574671846 for reading
==> amazon-ebs: Provisioning with shell script: /var/folders/yy/ff0tkhr141xdxvxf5fz43kmw0000gn/T/packer-shell574671846
2018/03/20 11:31:38 packer: 2018/03/20 11:31:38 [INFO] 30 bytes written for 'uploadData'
2018/03/20 11:31:38 packer: 2018/03/20 11:31:38 opening new ssh session
2018/03/20 11:31:38 [INFO] 30 bytes written for 'uploadData'
2018/03/20 11:31:38 packer: 2018/03/20 11:31:38 starting remote command: test -d "/tmp/script_7410.sh"
2018/03/20 11:31:48 packer: 2018/03/20 11:31:48 Retryable error: Error uploading script: Please login as the user "ec2-user" rather than the user "root".
2018/03/20 11:31:48 packer:
```

Which is more similar to behavior in packer 1.0.0

**Comment by mwhooker** [comment]
still need to test sftp and other general sanity testing. Also want to check the other places we use `test -d`

**Comment by SwampDragons** [comment]
Yeah, it turns out the `test -d` part is a bad assumption on my part; the ssh communicator could also be used for windows and in that case it'll have issues.  It hasn't totally broken ssh with windows, but it does mean that it behaves less predictably.

File: communicator/ssh/communicator.go [hunk 1]
```diff
@@ -585,7 +604,16 @@ func (c *comm) scpUploadSession(path string, input io.Reader, fi *os.FileInfo) e
...
+	if stdout.Len() > 0 {
+		return fmt.Errorf("%s", stdout.Bytes())
+	}
+	if stderr.Len() > 0 {
+		return fmt.Errorf("%s", stderr.Bytes())
+	}
 	if cmd.ExitStatus == 0 {
+		if fi == nil {
+			return fmt.Errorf("Upload path is a directory, unable to continue.")
+		}
```
Comment (SwampDragons) [inline communicator/ssh/communicator.go]: My brain is really really sluggish today. When is fi nil?

**Comment by SwampDragons** [comment]
This looks fine to me. We can probably look into fixing the assumptions around `test -d` here or in a different PR so as to not overload this one. Maybe we can access `provisioner/guest_commands.go` for that.

File: communicator/ssh/communicator.go (hunk 1 above)
Comment (mwhooker) [reply communicator/ssh/communicator.go]: more details in https://github.com/hashicorp/packer/issues/6032

**Comment by SwampDragons** [comment]
oh yeah this error occurs again in the sftp code: https://github.com/hashicorp/packer/blob/master/communicator/ssh/communicator.go#L427

**Comment by mwhooker** [comment]
fersher, but this PR fixes ;)

**Comment by shadycuz** [comment]
Thanks for the quick fixes =) 👍

**Comment by mwhooker** [comment]
We can't modify the destination path in the communicator because it breaks assumptions in the provisioners. For example, we try to chmod in the shell provisioner. The chmod command uses the path as supplied by the user. If the communicator decides to rewrite the path, the provisioner doesn't know that, and so tries to chmod the wrong thing.

**Comment by mwhooker** [comment]
using this to test

upload to directory:

```json
{
    "builders": [
        {
            "ami_name": "packer-qs-{{timestamp}}",
            "instance_type": "t2.micro",
            "region": "us-east-1",
            "source_ami_filter": {
                "filters": {
                    "name": "*ubuntu-xenial-16.04-amd64-server-*",
                    "root-device-type": "ebs",
                    "virtualization-type": "hvm"
                },
                "most_recent": true,
                "owners": [
                    "099720109477"
                ]
            },
            "ssh_username": "ubuntu",
            "ssh_file_transfer_method": "sftp",
            "type": "amazon-ebs"
        }
    ],
    "provisioners": [
        {
            "inline": [
                "sleep 3",
                "echo foo"
            ],
            "remote_path": "/tmp",
            "type": "shell"
        }
    ]
}
```

upload using wrong username:

```json
{
    "builders": [
        {
            "ami_name": "packer-qs-{{timestamp}}",
            "instance_type": "t2.micro",
            "region": "us-east-1",
            "source_ami": "ami-a4c7edb2",
            "ssh_pty": true,
            "ssh_file_transfer_method": "sftp",
            "ssh_username": "root",
            "type": "amazon-ebs"
        }
    ],
    "provisioners": [
        {
            "inline": [
                "echo foo"
            ],
            "type": "shell"
        }
    ]
}
```

**Comment by mwhooker** [comment]
I'm not sure why the change in provisioner/shell/provisioner.go didn't already exist. There may be something I'm not seeing. Will need to test a little bit.

Adding a file info struct allows the uploader to bypass an intermediate temp file and just copy directly from the file object. It lets the uploader set the correct remote file permissions. I can see this being extremely useful for the file provisioner, but maybe not for the shell provisioner, since we chmod it anyway. I might drop that commit. Will be sure to comment if I do.

I think there's some history in a ~18 month old PR that I haven't looked for yet.

**Comment by mwhooker** [comment]
I've changed this quite a bit and would like a re-review @SwampDragons. I unfortunately had to drop any directory detection cleverness we were doing. Reasons are explained in the commit that dropped it.

File: communicator/ssh/communicator.go [hunk 2]
```diff
@@ -413,18 +413,22 @@ func (c *comm) sftpUploadFile(path string, input io.Reader, client *sftp.Client,
...
 	if err != nil {
-		log.Printf("Unable to check whether remote path is a dir: %s", err)
+		log.Printf("[ERROR] Unable to check whether remote path is a dir: %s", err)
 		return err
 	}
 	cmd.Wait()
 	if cmd.ExitStatus == 0 {
-		log.Printf("path is a directory; copying file into directory.")
-		path = filepath.Join(path, filepath.Base((*fi).Name()))
+		return fmt.Errorf(
```
Comment (SwampDragons) [inline communicator/ssh/communicator.go]: I'm fine with this; it's explicit, which is the most important thing. I can change this for the other provisioners too, in a different PR, so they remain consistent.

**Comment by SwampDragons** [comment]
Which raises the question... maybe we _shouldn't_ be chmoding the shell provisioner?

**Pull request merged**